Accept github gists

lwronski opened the pull request.

alexarchambault commented on modules/build/src/main/scala/scala/build/Inputs.scala:
```diff
@@ -217,41 +222,50 @@ object Inputs {
     directories: Directories,
     baseProjectName: String,
-    download: String => Either[String, Array[Byte]],
+    download: String => Either[String, List[UrlPath]],
```
I think we shouldn't need to change that, and add `UrlPath`, here. The `Array[Byte]` could be changed to `os.Path` if needed, but that would only be a nice-to-have (see my comment below).

alexarchambault commented on modules/build/src/main/scala/scala/build/Inputs.scala:
```diff
@@ -217,41 +222,50 @@ object Inputs {
-                Option(u.getAuthority).iterator ++
-                Option(u.getPath).iterator.flatMap(_.split('/').iterator)
-              os.sub / it.filter(_.nonEmpty).toVector
+          download(arg).map { files =>
```
The Gist detection could be done around here. If a Gist URL is detected, it can be changed to the Gist zip URL (see the "Download ZIP" button on every Gist), and the zip file be downloaded here.

With the zip file, we can have a look at its content without unzipping it on disk, using the `java.util.zip.*` APIs (`ZipInputStream`, that can be passed a `ByteArrayInputStream` built from the `Array[Byte]`, or `ZipFile`, that can be passed a `File` if we change `download` to give us an `os.Path`)

This way, we can peek files from the Gist zip, and create `Virtual*` instances out of it, with a sensible `source` field (for `source`, we can try to build URLs like [this one](https://gist.github.com/alexarchambault/7b4ec20c4033690dd750ffd601e540ec#file-main-sc), pointing at the Gist individual files).

alexarchambault commented on modules/cli/src/main/scala/scala/cli/commands/SharedOptions.scala:
```diff
@@ -185,14 +186,36 @@ final case class SharedOptions(
         .left.map(_.describe)
-        .map(f => os.read.bytes(os.Path(f, Os.pwd)))
+        .map { f =>
+          // trying to unpack archive, support for downloading tar.gz files
```
We shouldn't need to unpack things on disk, see my comments above.

lwronski committed:
Accept Github Gists

lwronski committed:
Fix - use method only from java 8

alexarchambault committed:
Don't use commons-io

Not sure, but it might pull some more classes in the cli call graph,
which make native-image ask for too much memory on the CI Windows
worker.
